abci/src/codec.rs

```diff
@@ -0,0 +1,174 @@
+//! Encoding/decoding mechanisms for ABCI requests and responses.
+//!
+//! Implements the [Tendermint Socket Protocol][tsp].
+//!
+//! [tsp]: https://docs.tendermint.com/master/spec/abci/client-server.html#tsp
+
+use crate::Result;
+use bytes::{Buf, BufMut, BytesMut};
+use prost::Message;
+use std::io::{Read, Write};
+use std::marker::PhantomData;
+use tendermint_proto::abci::{Request, Response};
+
+/// The maximum number of bytes we expect in a varint. We use this to check if
+/// we're encountering a decoding error for a varint.
+pub const MAX_VARINT_LENGTH: usize = 16;
+
+/// The server receives incoming requests, and sends outgoing responses.
+pub type ServerCodec<S> = Codec<S, Request, Response>;
+
+#[cfg(feature = "client")]
+/// The client sends outgoing requests, and receives incoming responses.
+pub type ClientCodec<S> = Codec<S, Response, Request>;
+
+/// Allows for iteration over `S` to produce instances of `I`, as well as
+/// sending instances of `O`.
+///
+/// `S` must implement [`std::io::Read`] and [`std::io::Write`].
+pub struct Codec<S, I, O> {
+    stream: S,
+    // Long-running read buffer
+    read_buf: BytesMut,
+    // Fixed-length read window
+    read_window: Vec<u8>,
+    write_buf: BytesMut,
+    _incoming: PhantomData<I>,
+    _outgoing: PhantomData<O>,
+}
+
+impl<S, I, O> Codec<S, I, O>
+where
+    S: Read + Write,
+    I: Message + Default,
+    O: Message,
+{
+    /// Constructor.
+    pub fn new(stream: S, read_buf_size: usize) -> Self {
+        Self {
+            stream,
+            read_buf: BytesMut::new(),
+            read_window: vec![0_u8; read_buf_size],
+            write_buf: BytesMut::new(),
+            _incoming: Default::default(),
+            _outgoing: Default::default(),
+        }
+    }
+}
+
+// Iterating over a codec produces instances of `Result<I>`.
+impl<S, I, O> Iterator for Codec<S, I, O>
+where
+    S: Read,
+    I: Message + Default,
+{
+    type Item = Result<I>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        loop {
+            // Try to decode an incoming message from our buffer first
+            match decode_length_delimited::<I>(&mut self.read_buf) {
+                Ok(opt) => {
+                    if let Some(incoming) = opt {
+                        return Some(Ok(incoming));
+                    }
+                }
+                Err(e) => return Some(Err(e)),
```

Comment:
imho I find the following version of the code a bit easier to understand, but that's completely subjective so feel free to ignore.

```suggestion
                Ok(Some(incoming)) => return Some(Ok(incoming)),
                Err(e) => return Some(Err(e)),
                _ => (), // not enough data to decode a message, let's continue.
```